File: app/lib/models/campaign.dart
/// Funil de uma campanha (contagem de destinatários por status).
class CampaignFunnel {
  CampaignFunnel({
    this.total = 0,
    this.pending = 0,
    this.sent = 0,
    this.delivered = 0,
    this.read = 0,
    this.replied = 0,
    this.failed = 0,
    this.skipped = 0,
  });

  final int total;
  final int pending;
  final int sent;
  final int delivered;
  final int read;
  final int replied;
  final int failed;
  final int skipped;

  /// Acumulados do funil (cada destinatário tem UM status; aqui somamos "chegou até").
  int get reachedSent => sent + delivered + read + replied;
  int get reachedDelivered => delivered + read + replied;
  int get reachedRead => read + replied;
  int get done => total - pending;

  factory CampaignFunnel.fromJson(Map<String, dynamic> j) => CampaignFunnel(
        total: (j['total'] as num?)?.toInt() ?? 0,
        pending: (j['pending'] as num?)?.toInt() ?? 0,
        sent: (j['sent'] as num?)?.toInt() ?? 0,
        delivered: (j['delivered'] as num?)?.toInt() ?? 0,
        read: (j['read'] as num?)?.toInt() ?? 0,
        replied: (j['replied'] as num?)?.toInt() ?? 0,
        failed: (j['failed'] as num?)?.toInt() ?? 0,
        skipped: (j['skipped'] as num?)?.toInt() ?? 0,
      );
}

/// Uma campanha de disparo de template.
class Campaign {
  Campaign({
    required this.id,
    required this.name,
    required this.templateName,
    required this.templateLang,
    this.bodyText,
    required this.status,
    required this.scheduledAt,
    required this.ratePerMin,
    required this.createdAt,
    required this.funnel,
  });

  final String id;
  final String name;
  final String templateName;
  final String templateLang;
  final String? bodyText;
  final String status; // scheduled | running | paused | done | canceled
  final DateTime scheduledAt;
  final int ratePerMin;
  final DateTime createdAt;
  final CampaignFunnel funnel;

  String get statusLabel => switch (status) {
        'scheduled' => 'Agendada',
        'running' => 'Enviando',
        'paused' => 'Pausada',
        'done' => 'Concluída',
        'canceled' => 'Cancelada',
        _ => status,
      };

  bool get isActive => status == 'scheduled' || status == 'running' || status == 'paused';

  factory Campaign.fromJson(Map<String, dynamic> j) => Campaign(
        id: j['id'],
        name: j['name'] ?? '',
        templateName: j['template_name'] ?? '',
        templateLang: j['template_lang'] ?? 'pt_BR',
        bodyText: j['body_text'],
        status: j['status'] ?? 'scheduled',
        scheduledAt: DateTime.tryParse(j['scheduled_at'] ?? '')?.toLocal() ?? DateTime.now(),
        ratePerMin: (j['rate_per_min'] as num?)?.toInt() ?? 12,
        createdAt: DateTime.tryParse(j['created_at'] ?? '')?.toLocal() ?? DateTime.now(),
        funnel: CampaignFunnel.fromJson((j['funnel'] as Map?)?.cast<String, dynamic>() ?? const {}),
      );
}

/// Destinatário de uma campanha (linha do detalhe).
class CampaignRecipient {
  CampaignRecipient({
    required this.id,
    required this.phone,
    this.contactName,
    required this.status,
    this.error,
  });

  final String id;
  final String phone;
  final String? contactName;
  final String status;
  final String? error;

  String get displayName => (contactName != null && contactName!.isNotEmpty) ? contactName! : phone;

  String get statusLabel => switch (status) {
        'pending' => 'Na fila',
        'sent' => 'Enviada',
        'delivered' => 'Entregue',
        'read' => 'Lida',
        'replied' => 'Respondeu',
        'failed' => 'Falhou',
        'skipped' => 'Pulada',
        _ => status,
      };

  factory CampaignRecipient.fromJson(Map<String, dynamic> j) => CampaignRecipient(
        id: j['id'],
        phone: j['phone'] ?? '',
        contactName: j['contact_name'],
        status: j['status'] ?? 'pending',
        error: j['error'],
      );
}

File: backend/internal/models/campaign.go
package models

import "time"

// Status de uma campanha.
const (
	CampaignScheduled = "scheduled" // aguardando a hora de início
	CampaignRunning   = "running"   // worker enviando (no ritmo configurado)
	CampaignPaused    = "paused"
	CampaignDone      = "done"
	CampaignCanceled  = "canceled"
)

// Status de um destinatário (funil).
const (
	RecipientPending   = "pending"
	RecipientSent      = "sent"
	RecipientDelivered = "delivered"
	RecipientRead      = "read"
	RecipientReplied   = "replied"
	RecipientFailed    = "failed"
	RecipientSkipped   = "skipped"
)

// CampaignFunnel agrega os destinatários por status.
type CampaignFunnel struct {
	Total     int `json:"total"`
	Pending   int `json:"pending"`
	Sent      int `json:"sent"`
	Delivered int `json:"delivered"`
	Read      int `json:"read"`
	Replied   int `json:"replied"`
	Failed    int `json:"failed"`
	Skipped   int `json:"skipped"`
}

// Campaign é um disparo de template para uma audiência.
type Campaign struct {
	ID           string         `json:"id"`
	Name         string         `json:"name"`
	TemplateName string         `json:"template_name"`
	TemplateLang string         `json:"template_lang"`
	BodyText     *string        `json:"body_text,omitempty"`
	Status       string         `json:"status"`
	ScheduledAt  time.Time      `json:"scheduled_at"`
	RatePerMin   int            `json:"rate_per_min"`
	Params       []string       `json:"params,omitempty"`    // valores das variáveis {{1}}, {{2}}…
	ImageURL     *string        `json:"image_url,omitempty"` // foto (modelo com cabeçalho de imagem)
	// Público escolhido — guardado para reabrir a campanha ao COPIAR.
	Audience   string   `json:"audience,omitempty"` // all | groups | tag | manual
	GroupIDs   []string `json:"group_ids,omitempty"`
	TagID      *string  `json:"tag_id,omitempty"`
	ContactIDs []string `json:"contact_ids,omitempty"`

	CreatedBy *string        `json:"created_by,omitempty"`
	CreatedAt time.Time      `json:"created_at"`
	Funnel    CampaignFunnel `json:"funnel"`
}

// AudienceRef é o público escolhido (serializado em campaigns.audience_ref).
type AudienceRef struct {
	GroupIDs   []string `json:"group_ids,omitempty"`
	TagID      *string  `json:"tag_id,omitempty"`
	ContactIDs []string `json:"contact_ids,omitempty"`
}

// CreateCampaignRequest cria uma campanha com a audiência resolvida na hora.
type CreateCampaignRequest struct {
	Name         string     `json:"name" binding:"required,min=2"`
	TemplateName string     `json:"template_name" binding:"required"`
	TemplateLang string     `json:"template_lang"`
	BodyText     string     `json:"body_text"`
	Audience     string     `json:"audience" binding:"required,oneof=all groups tag manual"`
	GroupIDs     []string   `json:"group_ids"`   // audience=groups: 1..N grupos de contatos
	TagID        *string    `json:"tag_id"`      // audience=tag: conversas com esta etiqueta
	ContactIDs   []string   `json:"contact_ids"` // audience=manual
	ScheduledAt  *time.Time `json:"scheduled_at"` // null = começa agora
	RatePerMin   int        `json:"rate_per_min"` // default 12; 1..60
	Params       []string   `json:"params"`       // um valor por variável do modelo; {nome} = nome do contato
	ImageURL     string     `json:"image_url"`    // foto p/ modelo com cabeçalho de imagem (URL pública nossa)
	// Preenchido pelo service (não vem do cliente): categoria do modelo, que
	// define o preço cobrado pela Meta por mensagem entregue.
	TemplateCategory string `json:"-"`
}

// CampaignRecipient é um destinatário (linha do detalhe da campanha).
type CampaignRecipient struct {
	ID          string     `json:"id"`
	ContactID   string     `json:"contact_id"`
	ContactName *string    `json:"contact_name,omitempty"`
	Phone       string     `json:"phone"`
	Status      string     `json:"status"`
	Error       *string    `json:"error,omitempty"`
	SentAt      *time.Time `json:"sent_at,omitempty"`
	RepliedAt   *time.Time `json:"replied_at,omitempty"`
}
